compiler/src/iree/compiler/Dialect/Flow/Transforms/CollapseDimensions.cpp

```diff
@@ -154,101 +160,298 @@ static bool isEligibleForCollapse(linalg::GenericOp genericOp) {
+          Operation *definingOp = operand.getDefiningOp();
+          return !definingOp || slice.count(definingOp) == 0;
+        })) {
+      leafs.push_back(collapseShapeOp);
+    }
+  }
+
+  // 3. Clone the leaf `tensor.collapse_shape` ops outside the dispatch.
   OpBuilder::InsertionGuard g(rewriter);
-  rewriter.setInsertionPoint(regionOp->getParentOp());
+  rewriter.setInsertionPoint(dispatchOp);
+  for (auto reshapeOp : leafs) {
+    Operation *clonedOp = rewriter.clone(*reshapeOp.getOperation());
+    rewriter.replaceOp(reshapeOp, clonedOp->getResults());
+  }
 
-  auto maybeRegionOp = Flow::wrapOpInDispatchRegion(rewriter, newGenericOp);
-  if (failed(maybeRegionOp))
-    return failure();
+  // 4. From the yielded values find any that are produced by
+  //    `tensor.expand_shape` operation and move them out of the dispatch. For
+  //    this a new `DispatchRegionOp` is needed. For values that are yielded and
+  //    produced from `tensor.expand_shape`, the type of the result changes. The
+  //    dynamic dimensions of the result type also need to be updated.
+  SmallVector<Type> newReturnTypes;
+  SmallVector<Value> newDynamicDims;
+  SmallVector<Value> newYieldVals;
+  SmallVector<SmallVector<ReassociationIndices>> allReassociationIndices;
+  ValueRange dynamicDimsList = dispatchOp.getResultDims();
+  Location loc = dispatchOp.getLoc();
+  for (Value yieldedValue : returnOp->getOperands()) {
+    auto expandShapeOp = yieldedValue.getDefiningOp<tensor::ExpandShapeOp>();
+    if (!expandShapeOp) {
+      // 4a. Keep the same yield value if the producer is not a
+      // `tensor.expand_shape` op.
+      newReturnTypes.push_back(yieldedValue.getType());
+      newYieldVals.push_back(yieldedValue);
+      continue;
+    }
 
-  // Replace old regionOp with the result of collapse
-  rewriter.replaceOp(regionOp, collapseResults);
+    // 4b. The return type is same as the type of the source of the
+    // `tensor.expand_shape`.
+    RankedTensorType collapsedShapeType = expandShapeOp.getSrcType();
+    newReturnTypes.push_back(collapsedShapeType);
+    newYieldVals.push_back(expandShapeOp.getSrc());
+    SmallVector<ReassociationIndices> reassociation =
+        expandShapeOp.getReassociationIndices();
+    ArrayRef<int64_t> expandedShape = expandShapeOp.getResultType().getShape();
+
+    // 4c. Dynamic dims of the result shape is obtained by taking the static
+    //     shape + dynamic dims and collapsing them using the same reassociation
+    //     map as the `tensor.expand_shape`.
+    for (auto [index, shape] : llvm::enumerate(collapsedShapeType.getShape())) {
+      int64_t staticCollapsedShape = 1;
+      SmallVector<OpFoldResult> dynamicCollapsedDims;
+      for (auto collapsedDim : reassociation[index]) {
+        if (expandedShape[collapsedDim] == ShapedType::kDynamic) {
+          dynamicCollapsedDims.push_back(dynamicDimsList.front());
+          dynamicDimsList = dynamicDimsList.drop_front();
```

Comment:
It seems to me like it would be possible for multiple results to use the same dynamic dims. If so, is it still correct to "pop_front" here without refreshing the list each iteration of the outermost for-loop here?

Reply:
Can you give me an example of what you mean. Also I think the dynamic dims is per dynamic dimension per result. So there is no "collapsing of operands" happening there. That would make it very hard to reason about.

Reply:
> Also I think the dynamic dims is per dynamic dimension per result

Oh never mind then, this makes sense. I was worried about there somehow being reuse of dynamic dims.